### PR TITLE
[ConstraintElim] Check if second op implies first for And.

### PR DESCRIPTION
**Pull request by fhahn**
Generalize checkAndSecondOpImpliedByFirst to also check if the second operand implies the first.

### CODE REVIEW

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 1]
```diff
@@ -1365,16 +1365,24 @@ removeEntryFromStack(const StackEntry &E, ConstraintInfo &Info,
...
+                         SmallVectorImpl<StackEntry> &DFSInStack) {
 
   CmpInst::Predicate Pred;
   Value *A, *B;
   Instruction *And = CB.getContextInst();
-  if (!match(And->getOperand(0), m_ICmp(Pred, m_Value(A), m_Value(B))))
+  CmpInst *CmpToCheck = cast<CmpInst>(CB.getInstructionToSimplify());
+  unsigned OtherOpIdx = And->getOperand(0) == CmpToCheck ? 1 : 0;
+
+  if (OtherOpIdx != 0 && isa<SelectInst>(And))
```
Comment (nikic) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1382]: Is this due to a poison-propagation issue?

File: llvm/test/Transforms/ConstraintElimination/and-implied-by-operands.ll [hunk 2]
```diff
@@ -128,7 +128,7 @@ define i1 @test_same_cond_for_and_select_form(i8 %x) {
 ; CHECK-LABEL: @test_same_cond_for_and_select_form(
 ; CHECK-NEXT:  entry:
 ; CHECK-NEXT:    [[C_1:%.*]] = icmp ugt i8 [[X:%.*]], 10
-; CHECK-NEXT:    [[AND:%.*]] = select i1 [[C_1]], i1 true, i1 false
+; CHECK-NEXT:    [[AND:%.*]] = select i1 [[C_1]], i1 [[C_1]], i1 false
```
Comment (nikic) [inline llvm/test/Transforms/ConstraintElimination/and-implied-by-operands.ll:131]: Why did this regress?

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1385]: Yep, added a comment plus a TODO to check if the first op may be poison, thanks!

File: llvm/test/Transforms/ConstraintElimination/and-implied-by-operands.ll (hunk 2 above)
Comment (fhahn) [reply llvm/test/Transforms/ConstraintElimination/and-implied-by-operands.ll:131]: In this case, both conds are the same, and in the new code we will try `OtherOpIdx = 1` first, which then hits the early-exit condition for selects; Not sure if it is worth avoiding that and complicating the conditions.

**Review by nikic — APPROVED** [review]
LGTM

**Pull request merged**